diff --git a/flashinfer/prefill.py b/flashinfer/prefill.py
index 47d725c5..44f4d8ac 100755
--- a/flashinfer/prefill.py
+++ b/flashinfer/prefill.py
@@ -34,7 +34,7 @@ from .jit import (
 )
 from .cudnn import cudnn_batch_prefill_with_kv_cache
 from .page import block_sparse_indices_to_vector_sparse_offsets, get_seq_lens
-from .quantization import packbits, segment_packbits
+from .quantization import packbits, segment_packbits, _get_indptr_for_packed_mask
 from .utils import (
     log2e,
     FP4Tensor,
@@ -1697,6 +1697,10 @@ class BatchPrefillWithPagedKVCacheWrapper:
                 paged_kv_last_page_len,
                 page_size,
             )
+            
+        if packed_custom_mask is not None and custom_mask is None:
+            mask_indptr = _get_indptr_for_packed_mask(mask_indptr)
+        
         if packed_custom_mask is None and custom_mask is not None:
             # create packed custom mask from custom mask
             packed_custom_mask, mask_indptr = segment_packbits(
diff --git a/flashinfer/quantization.py b/flashinfer/quantization.py
index 810b1f2a..0dd06cdd 100644
--- a/flashinfer/quantization.py
+++ b/flashinfer/quantization.py
@@ -22,7 +22,6 @@ import torch
 from .jit.quantization import gen_quantization_module
 from .utils import register_custom_op, register_fake_op
 
-
 @functools.cache
 def get_quantization_module():
     return gen_quantization_module().build_and_load()
@@ -122,10 +121,7 @@ def segment_packbits(
     --------
     packbits
     """
-    seglen = indptr[1:] - indptr[:-1]
-    packed_len = (seglen + 7) // 8
-    indptr_new = torch.zeros(len(indptr), dtype=indptr.dtype, device=indptr.device)
-    indptr_new[1:] = torch.cumsum(packed_len, 0)
+    indptr_new = _get_indptr_for_packed_mask(indptr)
     output_nnzs = indptr_new[-1].item()
 
     device = x.device
@@ -134,3 +130,12 @@ def segment_packbits(
     y = torch.empty(output_nnzs, dtype=torch.uint8, device=device)
     get_quantization_module().segment_packbits(x, indptr, indptr_new, bitorder, y)
     return y, indptr_new
+
+def _get_indptr_for_packed_mask(
+    indptr: torch.Tensor,
+) -> torch.Tensor:
+    seglen = indptr[1:] - indptr[:-1]
+    packed_len = (seglen + 7) // 8
+    indptr_new = torch.zeros(len(indptr), dtype=indptr.dtype, device=indptr.device)
+    indptr_new[1:] = torch.cumsum(packed_len, 0)
+    return indptr_new
\ No newline at end of file
